Implement RawSocket.readMessage/writeMessage.

robert-ancell opened the pull request:
Fixes https://github.com/dart-lang/sdk/issues/46328

Some notes on the implementation:
- A specific case where this feature will be used is in [D-Bus](https://github.com/canonical/dbus.dart/issues/245).
- This on `RawSocket` and not `Socket`, since when receiving a message you need to decide if using `recv` or `recvmsg`.
- I haven't done any additional wrapping of the file descriptors - they are just left as integers for the API user to interpret. This might either involve passing them to some FFI code, or using some [new Dart constructors that can take file descriptors](https://github.com/canonical/dbus.dart/issues/245).
- I've only done an implementation for Linux, though other platforms (Android?) may also make use of these.

robert-ancell committed:
Implement RawSocket.readMessage/writeMessage.

Fixes https://github.com/dart-lang/sdk/issues/46328

mraleph commented:
Review will happen on the corresponding Gerrit CL https://dart-review.googlesource.com/c/sdk/+/205067

robert-ancell committed:
Drop MSAN workaround that is no longer required

robert-ancell committed:
Replace NULL with nullptr

robert-ancell committed:
Fix control message count

robert-ancell committed:
Reduce nesting by returning from a function quicker

robert-ancell committed:
Merge branch 'master' into socket-send-recv-msg

robert-ancell committed:
Read available amount of data instead of fixed size

robert-ancell committed:
Remove support for sending unix credentials.

There is not a current use case for this, and it can be added back later if necessary.

robert-ancell committed:
Use sizeof instead of constant

mit-mit commented:
Looks like the gerrit review for this is now https://dart-review.googlesource.com/c/sdk/+/212036
